src/bsoncxx/lib/bsoncxx/v1/oid.cpp

```diff
@@ -16,13 +16,172 @@
 
 //
 
+#include <bsoncxx/v1/exception.hpp>
+
+#include <cstring>
+#include <string>
+
+#include <bsoncxx/private/bson.hh> // <winsock.h> via <bson/bson-compat.h>
+#include <bsoncxx/private/immortal.hh>
 #include <bsoncxx/private/type_traits.hh>
 
 namespace bsoncxx {
 namespace v1 {
 
+using code = v1::oid::errc;
+
 static_assert(is_regular<oid>::value, "bsoncxx::v1::oid must be regular");
 static_assert(is_semitrivial<oid>::value, "bsoncxx::v1::oid must be semitrivial");
 
+constexpr std::size_t oid::k_oid_length;
+
+oid::oid() {
+    // Ensure the Winsock DLL is initialized prior to calling `gethostname` in `bsoncxx::v1::oid::oid()`:
+    //  - bson_oid_init -> bson_context_get_default -> ... -> _bson_context_init_random -> gethostname.
+    struct WSAGuard {
+#if defined(_WIN32)
+        ~WSAGuard() {
+            (void)WSACleanup();
+        }
+
+        WSAGuard(WSAGuard&&) = delete;
+        WSAGuard& operator=(WSAGuard&) = delete;
+        WSAGuard(WSAGuard const&) = delete;
+        WSAGuard& operator=(WSAGuard const&) = delete;
+
+        WSAGuard() {
+            WSADATA wsaData;
+            if (WSAStartup((MAKEWORD(2, 2)), &wsaData) != 0) {
+                throw v1::exception{
+                    WSAGetLastError(), std::system_category(), "WSAStartup() failed in bsoncxx::v1::oid::oid()"};
+            }
+        }
+#endif
+    };
+
+    bson_oid_t oid;
+    ((void)WSAGuard{}, bson_oid_init(&oid, nullptr));
+    std::memcpy(_bytes.data(), oid.bytes, sizeof(oid.bytes));
+}
+
+oid::oid(std::uint8_t const* bytes, std::size_t len) {
+    if (!bytes) {
+        throw v1::exception{code::null_bytes_ptr};
+    }
+
+    if (len != this->size()) {
+        throw v1::exception{code::invalid_length};
+    }
+
+    std::memcpy(_bytes.data(), bytes, _bytes.size());
+}
+
+oid::oid(v1::stdx::string_view str) {
+    if (str.empty()) {
+        throw v1::exception{code::empty_string};
+    }
```

Comment:
This check is [not present in v_noabi](https://github.com/mongodb/mongo-cxx-driver/blob/ad622035ec9c9ab352b42449d5fcbf426e8782e9/src/bsoncxx/lib/bsoncxx/v_noabi/bsoncxx/oid.cpp#L33), which may lead to library-level undefined behavior when invoking [`bson_oid_is_valid`](https://github.com/mongodb/mongo-c-driver/blob/d3165e022ae9155289e1910c72c393d035f75658/src/libbson/src/bson/bson-oid.c#L229).